### PR TITLE
feat: IAM-286 preliminary commit for charm base

### PR DESCRIPTION
**Pull request by wood-push-melon**
### What

This pull request aims to provide the preliminary version of the GLAuth base charm.

I'm sorry this pull request is chunky since it's based on the code used in the demo. Please focus on the code under `src` when reviewing.

### What's inside

- Basic juju event handling
- Postgresql charm relation
- Kubernetes ConfigMap for GLAuth configuration
- Integration with COS (sorry, this is because the code is based on the demo)

### Improvements on the way

- Kubernetes client and resource improvements for more extensibility
- Flaky scenario: remove integration and re-integrate with Postgresql charm (due to the varying delay mentioned in https://kubernetes.io/docs/concepts/configuration/configmap/#mounted-configmaps-are-updated-automatically). Already have an idea to improve this issue, and going to improve it in another pull request
- Pre-commit enforcement

### CODE REVIEW

**Review by nsklikas — COMMENTED** [review]
Overall looks good to me, but I have some comments. I suppose we are going to rework the `kubernetes_resource` logic either way (we should use the same logic in all of our charms).

File: src/validators.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    return wrapper
+
+
+def validate_database_resource(func: Callable):
+    @wraps(func)
+    def wrapper(self, *args: EventBase, **kwargs: Any):
+        event, *_ = args
+        logger.debug(f"Handling event: {event}")
+
+        self.unit.status = MaintenanceStatus("Configuring resources")
```
[the remark this answers is not in the document]
Comment (wood-push-melon) [reply src/validators.py:72]: Sorry, I am not sure why we need to remove this line.

File: src/charm.py [hunk 2]
```diff
@@ -1,103 +1,251 @@
...
+        self._configmap.delete()
+
+    def _on_database_created(self, event: DatabaseCreatedEvent):
+        self._update_glauth_config()
+        self._mount_glauth_config()
+        self._container.add_layer(
+            WORKLOAD_CONTAINER, self._pebble_layer, combine=True
+        )
+        self._restart_glauth_service()
+        self.unit.status = ActiveStatus()
```
[the remark this answers is not in the document]
Comment (wood-push-melon) [reply src/charm.py:206]: Yeah, definitely. I will mark it in the to-do list. We can try this new feature in another PR.

File: src/validators.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (nsklikas) [reply src/validators.py:72]: This is a validator, it is not configuring anything. No?

File: src/kubernetes_resource.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+class KubernetesResourceError(Exception):
+    def __init__(self, message: str):
+        self.message = message
+
+
+def watch_for_update(func: Callable):
```
Comment (nsklikas) [inline src/kubernetes_resource.py:29]: What is the use of this decorator?

File: src/validators.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+
+def validate_integration_exists(integration_name: str):
+    def decorator(func: Callable):
+        @wraps(func)
+        def wrapper(self: CharmBase, *args: EventBase, **kwargs: Any):
+            event, *_ = args
+            logger.debug(f"Handling event: {event}")
+
+            self.unit.status = MaintenanceStatus("Configuring resources")
```
Comment (nsklikas) [inline src/validators.py:49]: Correct me if I'm wrong, but you don't know if the wrapped function will actually configure resources. Perhaps the inner func should set this status if needed.

**Review by nsklikas — COMMENTED** [review]
One final comment. Also a lot of functions are missing return type annotations (and there are some other typing errors). Running `mypy` should show you whatever issues there are. It does not make sense to fix some of them, but let's try to resolve as many as possible.

**Review by nsklikas — APPROVED** [review]

**Pull request merged**